homeassistant/components/rainmachine/__init__.py

```diff
@@ -236,54 +231,20 @@ async def async_setup_entry(  # noqa: C901
     controller = get_client_controller(client)
 
     entry_updates: dict[str, Any] = {}
-    if not entry.unique_id or is_ip_address(entry.unique_id):
-        # If the config entry doesn't already have a unique ID, set one:
-        entry_updates["unique_id"] = controller.mac
+    update_controller_unique_id(entry, controller, entry_updates)
+    move_zone_run_time_to_options(entry, entry_updates)
 
-    if CONF_DEFAULT_ZONE_RUN_TIME in entry.data:
-        # If a zone run time exists in the config entry's data, pop it and move it to
-        # options:
-        data = {**entry.data}
-        entry_updates["data"] = data
-        entry_updates["options"] = {
-            **entry.options,
-            CONF_DEFAULT_ZONE_RUN_TIME: data.pop(CONF_DEFAULT_ZONE_RUN_TIME),
-        }
-    if CONF_USE_APP_RUN_TIMES not in entry.options:
-        entry_updates["options"] = {**entry.options, CONF_USE_APP_RUN_TIMES: False}
     if entry_updates:
-        hass.config_entries.async_update_entry(entry, **entry_updates)
+        update_config_entry(hass, entry, entry_updates)
 
-    if entry.unique_id and controller.mac != entry.unique_id:
-        # If the mac address of the device does not match the unique_id
-        # of the config entry, it likely means the DHCP lease has expired
-        # and the device has been assigned a new IP address. We need to
-        # wait for the next discovery to find the device at its new address
-        # and update the config entry so we do not mix up devices.
-        raise ConfigEntryNotReady(
-            f"Unexpected device found at {ip_address}; expected {entry.unique_id}, "
-            f"found {controller.mac}"
-        )
+    check_unexpected_device_discovery(hass, entry, controller, ip_address)
 
     async def async_update(api_category: str) -> dict:
         """Update the appropriate API data based on a category."""
         data: dict = {}
 
         try:
```

Comment:
Same here. Prefer it as it was before. Perhaps move the entire async_update function outside this function.